pinc/release_queue.inc

```diff
@@ -20,3 +25,322 @@ function cook_project_selector($project_selector)
+ *   * null - Don't filter out any queues
+ *   * string - Filter out queues whose name does not contain string
+ *   * array - Filter out queues whose name contains none of the strings
+ * @param null|string|array $group_selector
+ *   Select which groups and queues to return:
+ *   * null - Don't filter out any groups/queues.
+ *   * string - Filter out queues/groups whose group name does not contain string
+ *   * array - Filter out queues/groups whose group name contains none of the strings
+ * @param boolean show_groups
+ *   If true, return dummy group objects as well as queues
+ *
+ * @return generator
+ */
+function fetch_queues_data($round, $show, $show_groups, $name_selector = null, $group_selector = null)
+{
+    global $user_can_see_queue_settings;
+
+    // Building the WHERE clause is rather ad-hoc, mostly because
+    // a bunch of the output fields don't come directly from the DB result.
+    $proj_where = set_col_str("round_id", $round->id);
+    $clauses = [];
+    if (in_array($show, ["enabled", "populated"])) {
+        $clauses[] = "enabled = 1";
+    }
+    if (is_string($name_selector)) {
+        $name_selector = [$name_selector];
+    }
+    if (is_array($name_selector)) {
+        $values = array_map("DPDatabase::escape", $name_selector);
+        $clauses[] = surround_and_join($values, "name LIKE '%", "%'", ' OR ');
+    }
+    if (empty($clauses)) {
+        $clauses = ["1"];
+    }
+    $filter_where = surround_and_join($clauses, "(", ")", " AND ");
+    // We always want to see the rows that introduce groups.
+    $where = "$proj_where AND ($filter_where OR (SUBSTR(name, 1, 1) = '*'))";
+    $sql = "
+        SELECT *
+        FROM queue_defns
+        WHERE $where
+        ORDER BY ordering
+    ";
+
+    // Because queues are ordered, there is a convention to break them up
+    // into sections with a disabled queue at the start of the section
+    // indicating the section's name. These queue names start with "***"
+    // and we identify them here and treat them differently.
+    $group = null; // the most recently seen row defining a group
+    $res = DPDatabase::query($sql);
+    while ($row = mysqli_fetch_assoc($res)) {
+        if (startswith($row["name"], "***")) {
+            $group = str_replace("***", "", $row["name"]);
+            // Check for group
+            if ($show_groups && str_contains_any_of($group, $group_selector)) {
+                yield [
+                    "is_group" => true,
+                    "group" => $group,
+                ];
+            }
+            continue;
+        }
+
+        // Check for project in group.
+        if (!str_contains_any_of($group, $group_selector)) {
+            continue;
+        }
+
+        [$length, $unheld_length] = _get_queue_length($row["project_selector"], $round->project_waiting_state);
+        if ($show == "populated" && $length == 0) {
+            continue;
+        }
+        $data = _queue_data($row);
```

Comment:
Running `_get_queue_length()` against a whole lot of queues on a site with a whole lot of projects (eg: PROD) is non-trivial, so we should try not to do it twice for a single row -- here on line 155 and again in `_queue_data()`. It's probably worth special casing it, something like:
```php
function _queue_data($row, $length = null, $unheld_length = null)
{
    if ($length === null || $unheld_lenth === null) {
        [$length, $unheld_length] = ...
    }
```
and then here: `_queue_data($row, $length, $unheld_length)`

Reply:
Good idea! Done.